### PR TITLE
[Feature] MO AI : Vector BINARY IO

### PR DESCRIPTION
**Pull request by arjunsk**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [X] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue # https://github.com/matrixorigin/matrixone/issues/11458

## What this PR does / why we need it:

- Support Binary io for vectors. The textual insert of `array` is slower (300 inserts/sec), due to converting the array to a comma-separated string. So we are directly inserting `hex encoding of array` as a vector for faster performance (3000 inserts/sec) (more details [here](https://github.com/matrixorigin/docs/pull/213))
- Remove auto zero padding code for `vector` to prevent unexpected behaviors (e.g.:- b/b which should ideally give vector with all 1's will return division by zero error when b is zero-padded)
- Replace elementwise comparator with L2_Norm comparator for vectors as it is more appropriate for vector use case (Discussed with Dr. Min)
- Implement [encode() and decode()](https://www.postgresql.org/docs/9.0/functions-binarystring.html#FUNCTIONS-BINARYSTRING-OTHER) functions for inserting hex encoding of the vector from numpy.
- Optimized vector operations by removing unnecessary casting.
- Added a common error message for Vector related errors.

## Python Ingestion Code

<details>
<summary> Python Ingestion Code </summary>

```python
# CREATE DATABASE a;
# USE a;
# CREATE TABLE speedtest (id int, one_k_vector vecf32(1024));
import binascii
import time

import numpy as np
from sqlalchemy import create_engine, text
from sqlalchemy.orm import sessionmaker

table_name = "speedtest"
vec_len = 1024
num_inserts = 1024 * 8
num_vector_per_insert = 5


def run():
    engine = create_engine("mysql+pymysql://root:111@127.0.0.1:6001/a")
    Session = sessionmaker(bind=engine)
    session = Session()

    # mo macos vector dim=1024 vectors inserted=40960 insert/second=300 Text Insert
    # mo macos vector dim=1024 vectors inserted=40960 insert/second=3395 Hex Insert
    # mo macos vector dim=4048 vectors inserted=40960 insert/second=675  Hex Insert
    sql_insert = text("insert into speedtest (id, one_k_vector) values(:id, decode(:data,'hex') );")
    for i in range(num_inserts * num_vector_per_insert):
        arr = np.random.rand(vec_len)
        session.execute(sql_insert, {"id": i, "data": to_db_binary(arr), "vec_len": vec_len})
    session.commit()


def to_db_binary(value):
    if value is None:
        return value

    value = np.asarray(value, dtype='<f')  # little endian

    if value.ndim != 1:
        raise ValueError('expected ndim to be 1')

    return binascii.b2a_hex(value)


start = time.time()
run()
duration = time.time() - start
print(f"Result: vector dim={vec_len} vectors "
      f"inserted={num_inserts * num_vector_per_insert} "
      f"insert/second={num_inserts * num_vector_per_insert / duration}")


```

</details>

### SQL Binary Insert

```sql
mysql> create table t2(a int, b vecf32(3));

mysql>  insert into t2 values(2, decode("7e98b23e9e10383b2f41133f", "hex"));
     
mysql> select * from t2;
+------+---------------------------------------+
| a    | b                                     |
+------+---------------------------------------+
|    1 | [0.34881967, 0.0028086076, 0.5752134] |
+------+---------------------------------------+

mysql> select encode(b, "hex") from t2;
+--------------------------+
| encode(b, hex)           |
+--------------------------+
| 7e98b23e9e10383b2f41133f |
+--------------------------+

```

### CODE REVIEW

**Comment by matrix-meow** [comment]
@arjunsk Thanks for your contributions!

**Here are review comments for file pkg/common/moerr/error_no_ctx.go:**
## Review:

### Title
The title of the pull request is clear and concise. It indicates that the pull request is for a new feature related to MO AI and Vector Binary IO.

### Body
The body of the pull request provides a clear explanation of what the pull request does and why it is needed. It mentions the following changes:
- Support Binary insert and output for vectors to improve performance.
- Remove zero padding code for vectors to prevent unexpected behaviors.
- Replace elementwise comparator with L2_Norm comparator for vectors.

The body also includes Python ingestion code that demonstrates the usage of the new feature.

### Changes
The changes made in the pull request are focused on the file `pkg/common/moerr/error_no_ctx.go`. The changes add two new error functions `NewArrayInvalidOpNoCtx` and `NewArrayDefMismatchNoCtx` to handle errors related to vector operations and dimensions. The changes include the necessary error messages and formatting.

### Problems
1. Lack of documentation: The pull request does not include any documentation changes. It would be helpful to include documentation updates to explain the new feature and how to use it.

2. Inconsistent formatting: The changes made in the `error_no_ctx.go` file have inconsistent formatting. Some lines have an extra space before the opening parenthesis, while others do not. It would be better to have consistent formatting throughout the file.

### Suggestions
1. Documentation: Add documentation changes to explain the new feature and provide examples of how to use it. This will help users understand the feature and how it can be beneficial.

2. Consistent formatting: Ensure consistent formatting in the `error_no_ctx.go` file by removing the extra space before the opening parenthesis in all error functions.

### Optimization
The changes made in the pull request seem to be focused on adding new error functions and updating error messages. There may not be significant optimization opportunities in this specific pull request. However, it is always good to review the code for any potential performance improvements or code optimizations.

**Here are review comments for file pkg/container/types/array.go:**
## Review:

### Title
The title of the pull request is clear and concise. It indicates that the pull request is a feature enhancement related to MO AI and Vector Binary IO.

### Body
The body of the pull request provides a clear explanation of what the PR does and why it is needed. It mentions the following changes:
1. Support for binary insert and output for vectors to improve performance.
2. Removal of zero padding code for vectors to prevent unexpected behaviors.
3. Replacement of elementwise comparator with L2_Norm comparator for vectors.

The body also includes Python ingestion code for testing the changes.

### Changes
The changes made in the pull request are focused on the file `pkg/container/types/array.go`. Here are the specific changes made:

1. Added import statement for `encoding/hex` and `github.com/matrixorigin/matrixone/pkg/common/util`.
2. Added a new function `BlobToArray` that converts a blob (hex encoding of `[]T` array) to `[]T`.
3. Added a new function `ArrayToBlob` that converts `[]T` to hex encoding.
4. Updated the `BytesToArrayToString` function to use the `ArrayToString` function instead of `DecodeSlice`.
5. Added comments to the new functions.

### Problems to Address:

1. Lack of Documentation: The pull request does not include any documentation changes. It would be helpful to include documentation updates to explain the new functions and their usage.

2. Inconsistent Naming: The new functions `BlobToArray` and `ArrayToBlob` do not follow the naming convention used in the rest of the codebase. It would be better to use consistent naming, such as `HexToSlice` and `SliceToHex`.

3. Error Handling: The new functions `BlobToArray` and `ArrayToBlob` do not handle errors properly. It would be better to return an error instead of `nil` when an error occurs during decoding or encoding.

4. Code Optimization: The `BlobToArray` function uses `util.UnsafeBytesToString` to convert the input byte array to a string before decoding it. This conversion is unnecessary and can be optimized by directly passing the byte array to `hex.DecodeString`.

5. Code Duplication: The `BlobToArray` and `ArrayToBlob` functions have similar logic for decoding and encoding. It would be better to refactor the code to avoid duplication and improve maintainability.

### Suggestions to Address the Problems:

1. Documentation: Add documentation comments for the new functions `BlobToArray` and `ArrayToBlob` to explain their purpose, input parameters, and return values. Also, update the existing documentation to reflect the changes made in the pull request.

2. Naming: Rename the functions `BlobToArray` and `ArrayToBlob` to `HexToSlice` and `SliceToHex`, respectively, to follow the naming convention used in the codebase.

3. Error Handling: Update the `BlobToArray` and `ArrayToBlob` functions to return an error when an error occurs during decoding or encoding. This will provide better error handling and make the code more robust.

4. Code Optimization: Remove the unnecessary conversion from byte array to string in the `BlobToArray` function. Instead, directly pass the byte array to `hex.DecodeString`.

5. Code Duplication: Refactor the code to avoid duplication in the `BlobToArray` and `ArrayToBlob` functions. Extract the common logic for decoding and encoding into a separate function and reuse it in both functions.

### Additional Suggestions:

1. Unit Tests: Add unit tests to ensure the correctness of the new functions and to cover different scenarios.

2. Code Formatting: Ensure consistent code formatting throughout the file, including indentation and spacing.

3. Code Review: Consider reviewing the changes with other team members to get additional feedback and ensure the quality of the code.

4. Performance Testing: Conduct performance testing to validate the claim of improved performance for binary insert and output for vectors.

Overall, the pull request addresses an important feature enhancement related to MO AI and Vector Binary IO. However, there are some problems that need to be addressed, such as lack of documentation, inconsistent naming, error handling, code optimization, and code duplication. By addressing these problems and considering the additional suggestions, the quality and maintainability of the codebase can be improved.

**Here are review comments for file pkg/sql/plan/function/func_cast.go:**
Title: [Feature] MO AI : Vector BINARY IO

Body:
- The pull request adds support for binary insert and output for vectors.
- It removes zero padding code for vectors to prevent unexpected behaviors.
- It replaces the elementwise comparator with L2_Norm comparator for vectors.

Changes made in the file pkg/sql/plan/function/func_cast.go:
- Added support for casting from types T_array_float32 and T_array_float64 to types T_blob.
- Added a new function `blobToArray` to handle casting from BLOB to ARRAY for VECTOR FAST/BINARY IO.
- Modified the `strToStr` function to handle casting from VARCHAR to BLOB.
- Added a new function `ArrayToBlob` to handle casting from ARRAY to BLOB for VECTOR FAST/BINARY IO.
- Modified the `ArrayToArray` function to handle casting from ARRAY to ARRAY with different types.

Review:

1. The title of the pull request is clear and concise, indicating that it is a feature related to vector binary I/O for MO AI.

2. The body of the pull request provides a clear explanation of what the PR does and why it is needed. It mentions the addition of support for binary insert and output for vectors, the removal of zero padding code, and the replacement of the elementwise comparator with L2_Norm comparator. It also references the related issue #11458.

3. The changes made in the file pkg/sql/plan/function/func_cast.go seem to be related to adding support for casting from BLOB to ARRAY and from ARRAY to BLOB for VECTOR FAST/BINARY IO. The changes include adding new functions `blobToArray` and `ArrayToBlob`, modifying the `strToStr` and `ArrayToArray` functions, and updating the `supportedTypeCast` map.

Suggestions:

1. It would be helpful to provide more context or explanation in the body of the pull request regarding the motivation behind adding support for binary insert and output for vectors, as well as the decision to remove zero padding code and replace the elementwise comparator. This would help reviewers understand the reasoning behind these changes and provide more targeted feedback.

2. The changes made in the file pkg/sql/plan/function/func_cast.go could benefit from more detailed comments explaining the purpose and functionality of the new functions `blobToArray` and `ArrayToBlob`, as well as the modifications made to the existing functions. This would make it easier for future developers to understand and maintain the code.

3. It would be helpful to include some test cases or examples in the pull request to demonstrate the functionality and correctness of the changes made. This would provide more confidence in the implementation and make it easier for reviewers to verify the changes.

4. Consider optimizing the code by removing any unnecessary or redundant code. This could include

**Here are review comments for file pkg/sql/util/eval_expr_util.go:**
## Review:

### Title
The title of the pull request is clear and concise. It indicates that the pull request is for a new feature related to MO AI and Vector Binary IO.

### Body
The body of the pull request provides the following information:
- The type of the pull request (Improvement and Feature)
- The issue number that this pull request fixes
- The purpose of the pull request, which includes:
  - Support for binary insert and output for vectors
  - Removal of zero padding code for vectors to prevent unexpected behavior
  - Replacement of elementwise comparator with L2_Norm comparator for vectors
- Ingestion code in Python for testing the feature

### Changes
The changes made in the pull request are focused on the file `pkg/sql/util/eval_expr_util.go`. The changes include:
- Removing the zero padding code for vectors
- Adding checks for the length of the input array to match the destination length
- Removing the error message for source length being larger than destination length

### Problems
1. Lack of explanation for the removal of zero padding code: The pull request mentions the removal of zero padding code for vectors but does not provide any explanation for why this change is necessary. It would be helpful to include a brief explanation in the pull request body or as a comment in the code.
2. Inconsistent use of error handling: The changes made in the code include adding checks for the length of the input array to match the destination length, but the error message for source length being larger than destination length is removed. It is unclear why the error message is removed and if there are any alternative error handling mechanisms in place.
3. Lack of explanation for the replacement of elementwise comparator: The pull request mentions the replacement of elementwise comparator with L2_Norm comparator for vectors, but does not provide any explanation for why this change is necessary. It would be helpful to include a brief explanation in the pull request body or as a comment in the code.

### Suggestions
1. Provide an explanation for the removal of zero padding code: Add a comment in the code or update the pull request body to explain why the zero padding code for vectors is being removed. This will help reviewers and future maintainers understand the rationale behind the change.
2. Clarify the error handling mechanism: If the error message for source length being larger than destination length is removed intentionally, provide an explanation in the pull request body or as a comment in the code. If there are alternative error handling mechanisms in place, document them to ensure clarity.
3. Provide an explanation for the replacement of elementwise comparator: Add a comment in the code or update the pull request body to explain why the elementwise comparator is being replaced with the L2_Norm comparator for vectors. This will help reviewers and future maintainers understand the rationale behind the change.

### Optimization
The changes made in the pull request seem to be focused on improving the performance and behavior of vector binary IO. However, without further context or benchmarks, it is difficult to determine if the changes are optimized. It would be helpful to provide more information about the performance improvements achieved with these changes, such as benchmarks or comparisons with previous implementations.

Additionally, it would be beneficial to include unit tests for the new feature to ensure its correctness and to serve as documentation for future maintainers.

**Here are review comments for file pkg/vectorize/moarray/array.go:**
Review:

Title:
The title of the pull request is "Feature: MO AI - Vector BINARY IO". The title is clear and concise, indicating that the pull request is adding a new feature related to binary input/output for vectors.

Body:
The body of the pull request provides some context and explanation for the changes made. It mentions that the textual insert for vectors is slower and proposes using hex encoded binary for faster performance. It also mentions the removal of zero padding code for vectors to prevent unexpected behavior and the replacement of elementwise comparator with L2_Norm comparator. The body also includes Python code for ingestion and provides some performance results.

Changes:
The changes made in the pull request are focused on the file `pkg/vectorize/moarray/array.go`. Here are the main changes:

1. The constant `DimensionMismatchErrMsg` is removed. This constant was previously used to indicate dimension mismatch errors, but it is no longer needed.

2. The functions `Add`, `Subtract`, `Multiply`, and `Divide` now return an error of type `moerr.NewArrayInvalidOpNoCtx` instead of `moerr.NewInvalidInputNoCtx` when there is a dimension mismatch between the input vectors. This change improves the error handling and provides more specific error messages.

3. The function `Compare` is modified to use L2 norm comparison instead of elementwise comparison. This change improves the accuracy of the comparison and helps in ordering arrays by nearness on the cartesian plane.

4. The functions `L1Norm`, `L2Norm`, and `CosineSimilarity` now return an error of type `moerr.NewArrayInvalidOpNoCtx` when there is a dimension mismatch between the input vectors. This change improves the error handling and provides more specific error messages.

Suggestions:
1. In the body of the pull request, it would be helpful to provide more information about the performance improvement achieved by using hex encoded binary for vector input/output. This could include more detailed performance results and comparisons with the previous textual insert method.

2. The Python ingestion code provided in the body of the pull request could be moved to a separate file or section to improve readability.

3. In the changes made to the `pkg/vectorize/moarray/array.go` file, it would be helpful to include comments explaining the rationale behind the changes and the benefits they provide. This would make it easier for future developers to understand the code.

4. It would be beneficial to include unit tests for the modified functions in the `pkg/vectorize/moarray/array.go` file to ensure that the changes are working as expected and to provide a safety net for future changes.

5. Consider using more descriptive variable names in the `Compare` function to improve code readability.

Overall, the pull request introduces a useful feature for binary input/output for vectors and makes improvements to error handling and comparison methods. With some additional improvements and testing, the changes made in the pull request can be merged into the codebase.

**Here are review comments for file test/distributed/cases/array/array.result:**
Title: [Feature] MO AI : Vector BINARY IO

Body: This pull request adds support for binary insert and output for vectors in the MO AI project. The purpose of this feature is to improve performance by directly inserting hex encoded binary as vectors instead of using textual insertions. The pull request also removes zero padding code for vectors to prevent unexpected behaviors and replaces the elementwise comparator with L2_Norm comparator for vectors.

Changes Made:
- Added Python ingestion code for creating a database, table, and inserting vectors using binary insert.
- Modified the array.result file to include tests for vector operations and vector dimension mismatches.

Review:

1. Title: The title of the pull request is clear and concise, indicating that it is a feature addition related to vector binary IO in the MO AI project.

2. Body: The body of the pull request provides a clear explanation of what the PR does and why it is needed. It mentions the improvements made, such as faster performance with binary insertions and prevention of unexpected behaviors. It also references the related issue on GitHub.

3. Changes Made: The changes made include adding Python ingestion code for creating a database, table, and inserting vectors using binary insert. The array.result file is also modified to include tests for vector operations and vector dimension mismatches.

Suggestions for Improvement:

1. In the body of the pull request, it would be helpful to provide more details about the performance improvement achieved with binary insertions compared to textual insertions. This could include specific benchmarks or metrics.

2. The Python ingestion code could be further optimized by using prepared statements instead of executing the same SQL statement in a loop. This would improve performance by reducing the overhead of preparing and parsing the SQL statement for each iteration.

3. In the array.result file, the error messages for dimension mismatches could be more informative. Instead of just stating "invalid input," it would be helpful to include the expected and actual dimensions in the error message.

4. The array.result file could also include tests for other vector operations, such as subtraction, multiplication, and division, to ensure that they work correctly with binary insertions.

Overall, the pull request addresses the need for binary insert and output for vectors in the MO AI project and makes improvements to performance and behavior. However, there are some areas where further optimization and improvement can be made.

**Here are review comments for file test/distributed/cases/array/array.sql:**
## Review:

### Title
The title of the pull request is clear and concise. It indicates that the pull request is for a new feature related to MO AI and Vector BINARY IO.

### Body
The body of the pull request provides a clear explanation of what the PR does and why it is needed. It mentions the following changes:
1. Support for binary insert and output for vectors to improve performance.
2. Removal of zero padding code for vectors to prevent unexpected behaviors.
3. Replacement of elementwise comparator with L2_Norm comparator for vectors.

The body also includes a Python ingestion code snippet and a diff of changes made to the `array.sql` file.

### Changes
The changes made in the pull request include the following:
1. Addition of Python ingestion code for inserting vectors as binary.
2. Modification of the `array.sql` file to include tests for inserting vectors as binary and outputting vectors as binary.

### Problems
1. Lack of documentation: The pull request does not include any documentation changes. It would be helpful to include documentation updates to explain the new feature and how to use it.
2. Lack of test cases: The pull request does not include any new test cases to verify the functionality of the new feature. It would be beneficial to include test cases to ensure the correctness of the implementation.
3. Lack of code comments: The Python ingestion code snippet does not include any comments to explain the purpose of each section of code. Adding comments would improve the readability and maintainability of the code.
4. Inconsistent formatting: The code changes in the `array.sql` file have inconsistent formatting. It would be better to maintain consistent formatting throughout the file for better readability.

### Suggestions
1. Documentation: Add documentation changes to explain the new feature and how to use it. This will help users understand the purpose and functionality of the feature.
2. Test cases: Include new test cases to verify the functionality of the new feature. This will ensure that the feature works as expected and prevent any regressions.
3. Code comments: Add comments to the Python ingestion code snippet to explain the purpose of each section of code. This will make the code easier to understand and maintain.
4. Formatting: Ensure consistent formatting in the `array.sql` file. This will improve the readability of the code and make it easier to review and understand.

### Optimization
1. Performance improvements: Consider optimizing the code for better performance. This could include optimizing the binary insert and output operations to further improve the performance of the feature.
2. Error handling: Add proper error handling to handle any potential errors that may occur during the binary insert and output operations. This will make the code more robust and prevent unexpected failures.

Overall, the pull request introduces a new feature related to MO AI and Vector BINARY IO. However, there are some areas that need improvement, such as documentation, test cases, code comments, and formatting. Additionally, optimizing the code and adding error handling would further enhance the quality of the feature.

**Review by aressu1985 — APPROVED** [review]

**Review by fengttt — COMMENTED** [review]
cast maybe OK, not sure.     Why not use postgres encode function syntax?

encode(data, format) and decode(data, format)

where format can be 'hex', 'base64', etc.

**Comment by arjunsk** [comment]
> cast maybe OK, not sure. Why not use postgres encode function syntax?
> 
> encode(data, format) and decode(data, format)
> 
> where format can be 'hex', 'base64', etc.

Did you mean something like this

```sql
insert into t1 values( cast(decode('7e98b23e9e10383b2f41133f', 'hex') as vecf32(3)) );

select encode(b, 'hex') from t1;

```

**Comment by arjunsk** [comment]
Based on the review comments, I have replaced the `2 CAST` vector binary insert to `encode/decode`. Please review the new changes.

File: pkg/sql/plan/function/func_binary.go [hunk 1]
```diff
@@ -1789,6 +1791,55 @@ func EndsWith(ivecs []*vector.Vector, result vector.FunctionResultWrapper, proc
...
+		case "BASE64":
+			buf := make([]byte, base64.StdEncoding.DecodedLen(len(functionUtil.QuickBytesToStr(data))))
+			_, err = base64.StdEncoding.Decode(buf, data)
+			if err != nil {
+				return nil, err
+			}
+			return buf, nil
+			//TODO: Implement ESCAPE later.
+		default:
+			panic(moerr.NewInternalErrorNoCtx("unhandled format: %s", format))
```
Comment (m-schen) [inline pkg/sql/plan/function/func_binary.go:1838]: I'm not sure it should panic or just return an error.

**Review by m-schen — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

**Pull request merged**